File: src/main/java/ca/mcgill/science/tepid/client/CupsPrinterMgmt.java
package ca.mcgill.science.tepid.client;

import ca.mcgill.science.tepid.common.Utils;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;

public class CupsPrinterMgmt implements PrinterMgmt {

    private String user = null;

    @Override
    public boolean preBind() {
        return true; // Purposely blank
    }

    @Override
    public String tepidDataPath() {
        return System.getProperty("user.home") + "/.tepid";
    }

    @Override
    public void addPrinterImpl(String queue, String port, boolean isDefault) throws IOException, InterruptedException {
        if (user == null)
            user = Main.tokenUser == null || Main.tokenUser.isEmpty() ? System.getProperty("user.name") : Main.tokenUser;
        File tmpPpd = File.createTempFile("tepid", ".ppd");
        Files.copy(Utils.getResourceAsStream("XeroxWorkCentre7556.ppd"), tmpPpd.toPath(), StandardCopyOption.REPLACE_EXISTING);
        ProcessBuilder pb = new ProcessBuilder("sudo", "lpadmin", "-p", queue + "-" + user, "-E", "-v", "lpd://localhost:8515/" + port, "-P", tmpPpd.getAbsolutePath());
        pb.inheritIO();
        Process p = pb.start();
        p.waitFor();
        tmpPpd.delete();
    }

    @Override
    public void deletePrinterImpl(String queue, String port) throws IOException, InterruptedException {
        ProcessBuilder pb = new ProcessBuilder("sudo", "lpadmin", "-x", queue + "-" + user);
        pb.inheritIO();
        Process p = pb.start();
        p.waitFor();
    }

    @Override
    public void cleanPrinters() {
        // Purposely empty
    }

}

File: src/main/java/ca/mcgill/science/tepid/client/printers/CupsPrinterMgmt.kt
package ca.mcgill.science.tepid.client.printers

import ca.mcgill.science.tepid.client.utils.Config
import ca.mcgill.science.tepid.common.Utils
import ca.mcgill.science.tepid.utils.WithLogging
import java.io.File
import java.nio.file.Files
import java.nio.file.StandardCopyOption

class CupsPrinterMgmt : PrinterMgmt {

    private companion object : WithLogging()

    private val user = Config.USER_NAME

    override fun preBind(): Boolean = true // Purposely blank

    override fun tepidDataPath(): String = System.getProperty("user.home") + "/.tepid"

    override fun addPrinterImpl(queue: String, id: String, isDefault: Boolean) {
        val tmpPpd = File.createTempFile("tepid", ".ppd")
        try {
            Files.copy(Utils.getResourceAsStream("XeroxWorkCentre7556.ppd"), tmpPpd.toPath(), StandardCopyOption.REPLACE_EXISTING)
            val pb = ProcessBuilder("sudo", "lpadmin", "-p", queue + "-" + user, "-E", "-v", "lpd://localhost:8515/" + id, "-P", tmpPpd.absolutePath)
            pb.inheritIO()
            val p = pb.start()
            p.waitFor()
        } finally {
            tmpPpd.delete()
        }
    }

    override fun deletePrinterImpl(queue: String, id: String) {
        val pb = ProcessBuilder("sudo", "lpadmin", "-x", queue + "-" + user)
        pb.inheritIO()
        val p = pb.start()
        p.waitFor()
    }

    override fun cleanPrinters() = Unit

}
